### PR TITLE
add support for istio / envoy sidecar injection

### PR DESCRIPTION
**Pull request by montmanu**

### CODE REVIEW

File: DOCS.md [hunk 1]
```diff
@@ -19,6 +19,11 @@ The following parameters are used to configure this plugin:
 * `secrets` - variables to use in `secret_template`. These are base64 encoded by the plugin.
 * `secrets_base64` - variables to use in `secret_template`. These should already be base64 encoded; the plugin will not do so.
 
+[Istio](https://istio.io/) specific parameters:
+* `kube_inject` *optional*, _{bool}_ - if true, will run `istioctl kube-inject` on the generated kubernetes manifests (defaults to `false`)
```
Comment (jrynyt) [inline DOCS.md]: can you name all the options to indicate that they're isto-specific?

File: main.go [hunk 2]
```diff
@@ -295,6 +309,37 @@ func wrapMain() error {
...
+		for _, manifestPath := range pathArg {
+			manifestBase := filepath.Base(manifestPath)
+			manifestDir := filepath.Dir(manifestPath)
+			manifestExt := filepath.Ext(manifestPath)
+			manifestBaseWithoutExt := strings.Replace(manifestBase, manifestExt, "", 1)
+			// output path would resemble "/tmp/.kube.istio.yml"
+			istioOutputPath := filepath.Join(manifestDir, fmt.Sprintf("%s.istio%s", manifestBaseWithoutExt, manifestExt))
+			// use `--includeIPRanges`
+			if len(vargs.IncludeIPRanges) > 0 {
+				err = runner.Run(vargs.IstioctlCmd, "kube-inject", "--istioNamespace", vargs.IstioNamespace, "--includeIPRanges", vargs.IncludeIPRanges, "--filename", manifestPath, "--output", istioOutputPath)
```
Comment (jrynyt) [inline main.go]: can you define the command once, then conditionally add to it, to avoid duplication?

File: main.go [hunk 3]
```diff
@@ -295,6 +309,37 @@ func wrapMain() error {
...
+			}
+
+			if err != nil {
+				return fmt.Errorf("Error: %s\n", err)
+			}
+			istioInjectedPathArg = append(istioInjectedPathArg, istioOutputPath)
+		}
+
+		// Apply kube-inject'ed Kubernetes configuration files.
+		err = runner.Run(vargs.KubectlCmd, "apply", "--filename", strings.Join(istioInjectedPathArg, ","))
```
Comment (jrynyt) [inline main.go]: what exactly is being applied here that then requires the unmodified files to be applied below? I thought the injected files were augmented with the istio containers.

Comment (montmanu) [reply main.go]: ah. so, these aren't the unmodified files.. these are the modified files which are augmented with the istio containers

File: main.go (hunk 2 above)
Comment (montmanu) [reply main.go]: yeah, wasn't sure how execute a func in go with variadic arguments.. will track that down.

File: DOCS.md (hunk 1 above)
Comment (montmanu) [reply DOCS.md]: so, `istio_kube_inject` instead of `kube_inject`?

Comment (montmanu) [reply DOCS.md]: fwiw, they are currently named to align with the `istioctl kube-inject` [subcommand arguments](https://istio.io/docs/reference/commands/istioctl.html#istioctl-kube-inject) ...

**Comment by montmanu** [comment]
@jrynyt the NYTimes:0.4-istio-dev branch was created by @tonglil to isolate these istio-specific changes to a separate build of the plugin.. the `istioctl` API may be changing frequently, so we'd probably prefer to offer this feature as an opt-in, instead of bundling it with the mainline `drone-gke` plugin

**Comment by jrynyt** [comment]
ok, please make sure it doesn't get used in production.

**Comment by tonglil** [comment]
@montmanu can you add to the readme for this branch that this is experimental and a notice to not use it in production?

**Pull request merged**